File: 0289-game-of-life/0289-game-of-life.java
class Solution {
    public void gameOfLife(int[][] board) {
        int[][] directions = {{1,0}, {-1,0}, {0,1}, {0,-1}, {1,1}, {1,-1}, {-1,1}, {-1,-1}};
        List<int[]> deadCells = new ArrayList<>();
        List<int[]> liveCells = new ArrayList<>();

        for (int r = 0; r < board.length; r++) {
            for (int c = 0; c < board[0].length; c++) {
                int count = 0;

                for (int[] dir : directions) {
                    int newR = r + dir[0];
                    int newC = c + dir[1];
                    if (isBound(newR, newC, board)) {
                        count += board[newR][newC];
                    }
                }

                if (board[r][c] == 1) {
                    if (count < 2 || count > 3) {
                        deadCells.add(new int[]{r, c});
                    }
                } else {
                    if (count == 3) {
                        liveCells.add(new int[]{r, c});
                    }
                }
            }
        }

        for (int[] cell : deadCells) {
            board[cell[0]][cell[1]] = 0;
        }
        for (int[] cell : liveCells) {
            board[cell[0]][cell[1]] = 1;
        }
    }

    private boolean isBound(int r, int c, int[][] board) {
        return 0 <= r && r < board.length && 0 <= c && c < board[0].length;
    }
}

File: 0289-game-of-life/0289-game-of-life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        directions = [(1,0), (-1,0), (0,1), (0,-1), (1,1), (1,-1), (-1,1), (-1,-1)]
        deadcells = []
        livecells = [] 
        def isBound(r,c):
            return 0 <= r < len(board) and 0 <= c < len(board[0])

        for r in range(len(board)):
            for c in range(len(board[0])):
                count = 0

                for dr, dc in directions:
                    new_r = dr + r
                    new_c = dc+ c
                    if isBound(new_r, new_c):
                        count += board[new_r][new_c]

                if board[r][c] == 1:
                    if count < 2 or count > 3:
                        deadcells.append((r,c))
                else:
                    if count == 3:
                        livecells.append((r,c))

        for r,c in deadcells:
            board[r][c] = 0
        for r,c in livecells:
            board[r][c] = 1


